### DISCUSSION TITLE
Reuse form components on different forms

### DISCUSSION THREAD

**Discussion by guy-ecton** [mantinedev/mantine#7171] [Q&A]
When the examples use `useFormContext()` in a custom form input/component, it makes the component essentially tied to a specific form.
The other option to write a reusable custom form component is to pass the `form` as a prop.

I though I could do away with defining a generic `V` that specifies the current component's fields, but i get typescript errors:

```typescript
interface BirthdayFields {
  day: string | number
}

function Birthday<V extends {birthday: BirthdayFields}>({form}: {form: UseFormReturnType<V>}) {
 form.setFieldValue('birthday.day', 12)
//                                  ^ Argument of type 'number' is not assignable to parameter of type '"birthday" extends keyof V ? "day" extends keyof V[keyof V & "birthday"] ? V[keyof V & "birthday"][keyof V[keyof V & "birthday"] & "day"] : unknown : unknown'.
}
```
(the same code works when using `useFormContext()`)

So, how to smartly build reusable form components with mantine form?

Sources:
* https://mantine.dev/form/create-form-context/#store-context-in-separate-file
* https://mantine.dev/form/use-form/#useformreturntype
* [TS Playground](https://www.typescriptlang.org/play/?target=10#code/KYDwDg9gTgLgBDAnmYcAyEIGdgGliJYA8AagIYA2ArsFgHxwC8cA1gRAGZznW1wA+cABRYYUAJYA7AOZwAZHADeAXwCUAbgBQoSLATJUABTIwAFjxpEAKgBo4huKBjBJAEyzpMOfIWt0GzA5OLu5wAAYAJIpSHMBQcLjKAHRRMXFwAEq0MMphcAD8CY4gzm4ebIiccFYF9ibmlJZWANq4ALp2WaIMAFxwVJIskhAA7pJwfUElIeXsXDWFLYZtE-2Dw2NaOtDwSChwAMrAMABi4sAUrhbApI20AXBEZxeuxaWhGNh4BMTX9HRCMD1PrPS52ABud0m9Wut14WDsoNcdFUTAY4Ig4lcWj2qBO0AAtqQHopNHByXAcKdzpdrn0jtSXrCSHRNMpNJoOAMAMYwcQQcYcTCkN4zJQAI3EsFMrjIiD6ill8spYikskEkioBPFcWUygBQqgBJBhOJqNJFLghoJSSpSOuQgA5JLpUqkkrHXYAIwAJgAzKo2RygA)

**Comment by ryuhangyeong** [comment]
#7195

Is the question similar to this question?

**Reply by guy-ecton** [reply]
looks like it

**Comment by rtivital** [comment]
There is no good way of implementing this, you can either use `UseFormReturnType<any>` or make decision not to use such fields

**Comment by guy-ecton** [comment]
Yes, I ended up using `any` in the generic components (like Wizard<T>...) and having more concrete typing on the inputs themselves.

**Comment by jvllmr** [comment]
A similar pattern is now possible with Mantine v9. See: https://github.com/mantinedev/mantine/pull/8198
